### ISSUE TITLE
Custom playback positions

### ISSUE THREAD

**Issue by ThomS8312** [mpv-player/mpv#1171]
In VLC, they are called "custom bookmarks" and are like virtual chapters. Custom bookmarks can also have descriptions that correspond to them. They are very useful for navigating through lengthy videos and easy to modify. VLC saves custom bookmarks as a playlist to .xspf files to accomplish this, but mpv doesn't support XML playlist formats. Even so, that fact might not matter. VLC's implementation is quite buggy and maybe a simpler approach can be taken with mpv. Can mpv be made to do something similar?

**Comment by wm4** [comment]
Actually, there's this thing named ``ffmetadata``: https://www.ffmpeg.org/ffmpeg-formats.html#Metadata-1

In combination with ``--chapters-file``, this should give you the thing you requested.

**Comment by ThomS8312** [comment]
@wm4
Haven't had much time to get to this, but I briefly looked into <code>ffmetadata</code>. I couldn't get it to work with what I tried, and unsure of a few things.

1. What's <code>TIMEBASE</code> in layman's terms?
2. Can this be done with absolute time instead of seconds?
3. What's <code>line</code> in the <code>[STREAM]</code> section?

@nokangaroo
Thanks for the input. It looks a bit complicated, but I'll look into this if the simpler <code>ffmetadata</code> isn't robust enough.

**Comment by wm4** [comment]
>What's TIMEBASE in layman's terms?

It's the resolution of the time values. The example on the ffmpeg website uses ``TIMEBASE=1/1000``, which means 1 tick is 1 millisecond (1 second divided by 1000). The chapter end time is given as ``END=60000``, which with that timebase means 60000 milliseconds (= 60 seconds = 1 minute).

Yeah, this is a bit technical. Maybe it could be improved.

**Comment by nokangaroo** [comment]
Extracting an ffmetadata file with ffmpeg:
```
ffmpeg -i /file/with/metadata -f ffmetadata /path/to/ffmetadatafile
```
But what if the file does not have any metadata?

What's needed is an interface to do it on-the-fly: Read current position and
write it into a file. And present an interface to choose between user-set
positions (arbitrary positions, not DVD or mp4 chapters). My example code does
just that.

**Comment by nokangaroo** [comment]
Dig this. Is this really so complicated? Judge for yourself:
```
#!/bin/bash
# The player script for starting mpv with an environment

# Media folder:
MEDIADIR="$HOME/mpv-data"
[[ -d $MEDIADIR ]] || mkdir -p $MEDIADIR

#Input fifo:
FIFO=mpv.input
[[ -p $FIFO ]] || mkfifo $FIFO

# Identifying the media file (this is probably better than using a path)
# (I don't know yet if this is unique enough; head -c 2048 is too short)
MEDIA_ID=`head -c 4096 "$1" | md5sum | cut -d \  -f 1`
echo "MEDIA_ID=$MEDIADIR/$MEDIA_ID" > $MEDIADIR/data
echo "FIFO=$FIFO" >> $MEDIADIR/data

mpv --input-file=mpv.input "$1"
```

For more functionality, see my example DVD code:
```
#!/bin/bash
# The bookmarking script, to be used while mpv is running:
[[ `pidof mpv` ]] || exit

# Media folder:
MEDIADIR="$HOME/mpv-data"

# The file to source for variables:
DATAFILE=${MEDIADIR}/data

# Get variables:
[[ -f $DATAFILE ]] && . $DATAFILE

case $1 in
-c|--create)
# using an ffmetadata file would be very complicated, because you would have to
# know the start of the next bookmark (to mark the end of the present bookmark)
# before you even create it. You would have to create all bookmarks at one go,
# in sequence. The ffmetadata file format sucks.
BMK=$(zenity --entry --title="New Bookmark" --text="A new bookmark will be created at the present position. Choose a name for it:")
if [[ $BMK ]]; then
    #the bookmarks are numbered so they can be selected with the keypad:
    echo -e "$(expr `grep -s -o ^[0-9]* $MEDIA_ID | sort -n | tail -n 1` + 1)"'\t'"$BMK" >> $MEDIA_ID
    echo -n 'time=' >> $MEDIA_ID
    echo 'run "/bin/sh" "-c" "/usr/bin/echo ${time-pos} >> '$MEDIA_ID\" > $FIFO
fi
;;
-l|--list)
    SELECTION=`sed -n '/^[0-9]\+\t/ P' $MEDIA_ID | zenity --list --title="Bookmarks" --column="" --hide-header | cut -f 1`
    if [[ $SELECTION ]]; then
        timepos=`grep -s -w -A 1 "^$SELECTION" $MEDIA_ID | tail -n 1 | cut -d \= -f 2`
        echo "set time-pos $timepos" > $FIFO
    fi
;;
*)
#quit mpv
echo 'quit' > $FIFO
;;
esac
```

I tested this for, like, 5 minutes, so feel free to correct any errors.

**Comment by nokangaroo** [comment]
An improvement of the bookmarking script to handle arbitrary commands, not just
seeks:
```
#!/bin/bash
# Copyright (C) 2015 nokangaroo nokangaroo@NOSPAM.aon.at

# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.



# To be used while mpv, started with the player script, is running:
[[ `pidof mpv` ]] || exit

# Media folder:
MEDIADIR="$HOME/mpv-data"

# The file to source for variables:
DATAFILE=${MEDIADIR}/data

# Get variables:
[[ -f $DATAFILE ]] && . $DATAFILE

# Functions go here:
mpcmd() {
    echo "$*" > $FIFO
}

set_bookmark() {
    # This function can set arbitrary commands, not just seek times
    echo -e "$(expr `grep -s -o ^[0-9]* $MEDIA_ID | sort -n | tail -n 1` + 1)"'\t'"$BMK" >> $MEDIA_ID
    while [[ $# -gt 0 ]]; do
        echo -n 'command@'"$1 " >> $MEDIA_ID
        shift
        echo 'run "/bin/sh" "-c" "/usr/bin/echo \"${'"$1"'}\" >> '$MEDIA_ID\" > $FIFO
        sleep 0.3 #may or may not be needed
        shift
    done
    echo >> $MEDIA_ID #Newline is required for --list command to work
}
# End functions

case $1 in
-c|--create)
BMK=$(zenity --entry --title="New Bookmark" --text="A new bookmark will be created at the present position. Choose a name for it:")
if [[ $BMK ]]; then
    #the bookmarks are numbered so they can be selected with the keypad:
    set_bookmark "set time-pos" "=time-pos"
fi
;;
-k|--command)
# TODO: Enable graphical entry! (Difficult; zenity sucks)
# The distinction between bookmarks and commands is for convenience; they work
# in the same way
# An even number of arguments after the bookmark name is required. Arguments are
# evaluated in pairs with each command in a new line, see set_bookmark function
# The command can of course be combined with a seek
# Example: Set volume, audio and subtitle track as desired and do:
# bookmarkscript -k 'Name' 'set aid' '=aid' 'set sid' '=sid' 'set volume' '=volume'
# Repeat this for all tracks you want to switch between
BMK="$2"
if [[ "$3" && "$4" ]]; then
    shift 2
    set_bookmark "$@"
fi
;;
-l|--list)
# This can execute an arbitrary number of command@ lines after the bookmark:
SELECTION=`sed -n '/^[0-9]\+\t/ P' $MEDIA_ID | zenity --list --title="Bookmarks" --column="" --hide-header | cut -f 1`
if [[ $SELECTION ]]; then
    n=1
    while :; do
        command=`grep -s -w -A $n "^$SELECTION" $MEDIA_ID | tail -n 1 | cut -d \@ -f 2`
        [[ -n $command ]] && mpcmd "$command" || break
        n=`expr $n + 1`
    done
fi
;;
*)
#quit mpv
echo 'quit' > $FIFO
;;
esac
```
A sample bookmark file (the bookmarks should of course be given meaningful
names):
```
1	Track 1
command@set aid 1
command@set sid no
command@set volume 34.000000

2	bmk 1
command@set time-pos 33.424156

3	Track 2
command@set aid 2
command@set sid 1
command@set volume 47.000000

4	bmk 2
command@set time-pos 12.144156

5	bmk 3
command@set time-pos 3.144156
```

**Comment by ThomS8312** [comment]
Maybe a frontend can be created for <code>ffmetadata</code>? If it has the desired functionality, a user friendly frontend might work well for this.

**Comment by thebombzen** [comment]
There is now a way bookmark chapters and save them to an XML chapters file in the User Scripts Repository.  See [mpv-createchapter](https://github.com/shinchiro/mpv-createchapter).

**Comment by agreselin** [comment]
FWIW I too would like to have a quick way to make and edit chapters quickly. My use case is guitar tuturials, where the granularity of the splitting changes as I learn the song. So I'd be making and deleting chapters relatively often wrt to splitting a movie into parts. I really like mpv, this is the only thing I'm missing at the moment.

I've tried [mpv-bookmarker](https://github.com/NurioHin/mpv-bookmarker) but that's slower to navigate (it can only jump to the latest bookmark with a single key stroke) and most importantly its bookmarks are global, not a per-file thing.

@thebombzen [mpv-createchapter](https://github.com/shinchiro/mpv-createchapter) seems great but is there a way to make use of the exported file without burning the chapters onto the video?

**Comment by thebombzen** [comment]
At the moment the script doesn't support writing to other formats.

**Comment by Hrxn** [comment]
You could try an EDL-based playlist, I think..

**Comment by agreselin** [comment]
> You could try an [EDL](https://github.com/mpv-player/mpv/blob/master/DOCS/edl-mpv.rst)-based playlist, I think..

I've written a very simple [Awk script](https://gitlab.com/agreselin/scripts/-/blob/master/txt2edl) to convert a file of timestamps to an EDL file. It's not the ultimate in versatility but it covers my needs and maybe it can cover someone else's too.

Do you know of a way to auto-load it when I open a video, like with subtitles (is there a setting or an extension to use)?

**Comment by Hrxn** [comment]
Why not open the EDL file directly?

**Comment by agreselin** [comment]
Because I didn't know I could 😅. Still, is there a way to auto-load it on launching the video from the file manager?

**Comment by thebombzen** [comment]
put this atop your EDL file, and set the file to be executable. Then you can run it as a script: `#!/usr/bin/env -S sh -c 'tail -n +2 "$0" | mpv -'`.

**Comment by agreselin** [comment]
I was looking for something a bit different. In the file manager often I look for a certain thumbnail and play that file, more than going by file name. If there's an EDL with the same name as the video file (except for the extension) I'd like mpv to load it automatically, like it does for srt subtitles.

Still, interesting use of the shebang, I learned something :-)
